File: sample.java
// # Stack-1

// ## Problem1 Daily Temperatures (https://leetcode.com/problems/daily-temperatures/)

// Time Complexity :O(n)
// Space Complexity :O(n)
// Did this code successfully run on Leetcode :yes
// Any problem you faced while coding this :I didnt face any problem while coding this


// Your code here along with comments explaining your approach

class Solution {
    public int[] dailyTemperatures(int[] temperatures) {
        Stack<Integer> st = new Stack<>();
        int[] result = new int[temperatures.length];
        for(int i=0; i<temperatures.length; i++){
            while(!st.empty() && temperatures[i]>temperatures[st.peek()]){
            int popped = st.pop();
            result[popped] = i - popped;
        }st.push(i);
        }
        return result; 
    }
}

// ## Problem2 Next Greater Element II (https://leetcode.com/problems/next-greater-element-ii/)

// Time Complexity :O(n)
// Space Complexity :O(n)
// Did this code successfully run on Leetcode :yes
// Any problem you faced while coding this :I didnt face any problem while coding this


// Your code here along with comments explaining your approach

class Solution {
    public int[] nextGreaterElements(int[] nums) {
        int[] result = new int[nums.length];
        int n = nums.length;
        Stack<Integer> st = new Stack<>();
        Arrays.fill(result, -1);
        for(int i=0; i< 2 * n; i++){
            while(!st.isEmpty() && nums[i%n] > nums[st.peek()]){
                int popped = st.pop();
                result[popped] = nums[i%n];
            }
            if(i < n){
                st.push(i);
            }
        }
        return result;
    }
}

File: sample.swift
// Time Complexity : 0(n)
// Space Complexity : 0(n)
// Did this code successfully run on Leetcode :
// Any problem you faced while coding this :


// Your code here along with comments explaining your approach

    func nextGreaterElements(_ nums: [Int]) -> [Int] {
        if nums.count == 0 || nums == nil {
            return nums
        }
        var stack = [Int]()
        var output = Array(repeating: -1, count: nums.count)
        stack.append(0)
        for i in 0..<nums.count * 2 {
            var index = i % nums.count
            while !stack.isEmpty && nums[stack.last!] < nums[index] {
                output[stack.removeLast()] = nums[index]
            }
            if output[index] == -1 {
            stack.append(index)
            }
        }
        return output
    }